### PR TITLE
Develop

### PR DESCRIPTION
**Pull request by BrixInMotion**
After adding some new features and doing lots of improvements, it's time to do the next release.
New features are:

* Bidirectional mode for two DC motors
* Fully tested support for hall-sensor BLDC motors
  * supports normal mode and field weakening range
  * integrated PI regulator to set accurate RpM speed with changeable parameters
  * Tested on Arduino UNO and XMC4700
* Test sketch, to find the amount of pole pairs of any brushless motor

### CODE REVIEW

File: examples/BLDC_Find_Polepairs/BLDC_Find_Polepairs.ino [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+//Create an instance of 'IFX007TMotorControl' called 'MyMotor'
+IFX007TMotorControl MyMotor = IFX007TMotorControl();
+
+void setup()
+{
+  Serial.begin(115200);
+  Serial.println(" Infineon BLDC Polepairfinder ");
+  Serial.println(" ");
+  MyMotor.begin();
+  MyMotor.MotorParam.SensingMode = Hallsensor;
```
Comment (djaumann) [inline examples/BLDC_Find_Polepairs/BLDC_Find_Polepairs.ino:28]: I would do this as a define to make it clearer

File: examples/BLDC_Find_Polepairs/BLDC_Find_Polepairs.ino [hunk 2]
```diff
@@ -0,0 +1,96 @@
+/**
+ * This example code provides a test environment, to find out the number of polepairs of your BLDC motor.
+*/
+
+#include "IFX007T-Motor-Control.h"
+
+// ----------- Change to 0 if your motor has no hallsensor -------------
+bool Hallsensor = 1;
```
Comment (djaumann) [inline examples/BLDC_Find_Polepairs/BLDC_Find_Polepairs.ino]: I would do this as a define to make it clearer

File: examples/BLDC_HALL/BLDC_HALL.ino [hunk 3]
```diff
@@ -0,0 +1,42 @@
+/**
+ * This example code is for BLDC motor control with Hall Sensor.
+ * Please be noticed it has to be modified for a diffenrent motor (e.g. motor with differnet pole number, PI-values, ...)
+ * Refer to the README.md in this folder.
+*/
+
+#include "IFX007T-Motor-Control.h"
+uint16_t rpmSpeed = 800;
+bool direction = 0;           // 0 or 1
+bool weakening = 0;           // 0=normal, 1=weak
```
Comment (djaumann) [inline examples/BLDC_HALL/BLDC_HALL.ino]: Würde ich auf wieder als #defines machen

File: src/IFX007T-Motor-Control.cpp [hunk 4]
```diff
@@ -659,7 +895,7 @@ void IFX007TMotorControl::setADCspeedFast(void)
 
 void IFX007TMotorControl::setPwmFrequency(uint8_t pin, uint16_t divisor)
 {
-  setAnalogWriteFrequency(pin, 15000);       // 31250 Hz, obviously we nee the half (fault in XMCforArduino?)
+  setAnalogWriteFrequency(pin, 30000);       // 31250 Hz, obviously we nee the half (fault in XMCforArduino?)
```
Comment (djaumann) [inline src/IFX007T-Motor-Control.cpp:929]: Ja habe ich auch schon festgestellt, dass da was nicht passt. Ich denke das ist irgendein Pre-Scaler nicht richtig eingestellt. Ist mal wieder ein XMC-For-Arduino Fehler :)

File: src/IFX007T-Motor-Control.cpp [hunk 5]
```diff
@@ -165,27 +231,38 @@ void IFX007TMotorControl::configureBLDCMotor(BLDCParameter MyParameters)
     _debugPin = 1;
     
 
-    if(MyParameters.SensingMode)    //Hall-sensor mode (Not supported yet)
+    if(MyParameters.SensingMode)    //Hall-sensor mode
     {
         #define HALLmode
```
Comment (djaumann) [inline src/IFX007T-Motor-Control.cpp]: Defines immer alles GROSS schreiben z.B. #define KILIAN "Ist die Nummer 1"

File: src/IFX007T-Motor-Control.cpp [hunk 6]
```diff
@@ -165,27 +231,38 @@ void IFX007TMotorControl::configureBLDCMotor(BLDCParameter MyParameters)
...
+        DEBUG_PRINT_LN("Hallsensor mode");
+        _CurrentDutyCycle = 80;     // dutycycle at the beginning
+
+        //pinMode(_PinAssignment[AdcPin][0], INPUT_PULLUP);
+        //pinMode(_PinAssignment[AdcPin][1], INPUT_PULLUP);
+        //pinMode(_PinAssignment[AdcPin][2], INPUT_PULLUP);
+
+        // Enable interrupts for hallsensor inputs
+        //PCICR  = 0b00000010;               // Enable pin change interrupt for pins 14:8 (See Atmel 328 datasheet, page 79)
+        //PCMSK1 = 0b00001110;               // Enable pin change interrupt for pins A1 (PCINT9), A2 (PCINT10), A3 (PCINT11)
```
Comment (djaumann) [inline src/IFX007T-Motor-Control.cpp]: Wieso ist das alles auskommentiert?
Wenn das nicht gebraucht wird schmeiß es raus.

**Review by djaumann — COMMENTED** [review]
Soweit schaut es für mich sehr gut aus :) Über die Funktionalität kann ich jetzt nicht so viel sagen, aber da vertrau ich dir 👍 . Ansonsten die paar Kommentare die ich hatte sind jetzt nicht wirklich was.

Gute Arbeit Herr Brixel! 🥇

File: examples/BLDC_Find_Polepairs/BLDC_Find_Polepairs.ino (hunk 2 above)
Comment (BrixInMotion) [reply examples/BLDC_Find_Polepairs/BLDC_Find_Polepairs.ino]: Yes right, I changed it.

File: examples/BLDC_HALL/BLDC_HALL.ino (hunk 3 above)
Comment (BrixInMotion) [reply examples/BLDC_HALL/BLDC_HALL.ino]: Changed as well

File: src/IFX007T-Motor-Control.cpp (hunk 4 above)
Comment (BrixInMotion) [reply src/IFX007T-Motor-Control.cpp:929]: Ja das ist mal so und mal so

File: src/IFX007T-Motor-Control.cpp (hunk 5 above)
Comment (BrixInMotion) [reply src/IFX007T-Motor-Control.cpp]: Good to know, I changed the defines for the pin assignment as well to capital letters 👍

File: src/IFX007T-Motor-Control.cpp (hunk 6 above)
Comment (BrixInMotion) [reply src/IFX007T-Motor-Control.cpp]: You're right, it's a leftover of trying to implement an interrupt mode. Maybe relevant for the next release, but for now I deleted it.

File: src/IFX007T-Motor-Control.h [hunk 7]
```diff
@@ -65,21 +67,21 @@ class IFX007TMotorControl
...
 
-        void    setUniDirMotorSpeed(uint8_t motor, uint8_t dutycycle);          //For Unidirectional motors; Parameters: motor can be 0, 1 or 2, dutycycle can be 0 - 255
-        void    setBiDirMotorSpeed(bool direction, uint8_t dutycycle);          //For Bidirectional motors; Parametrs: direction can be 0 or 1, dutycycle can be 0 - 255
+        void    setUniDirMotorSpeed(uint8_t motor, uint8_t dutycycle);                      //For Unidirectional motors
+        void    setBiDirMotorSpeed(uint8_t motor, bool direction, uint8_t dutycycle);       //For Bidirectional motors
         void    configureBLDCMotor(BLDCParameter MyParameters);  
         void    setBLDCmotorRPMspeed(bool direction, uint16_t desired_rpmSpeed);
-        void    setBLDCDutyCyclespeed(bool direction, uint8_t dutycycle);                         
+        void    setBLDCDutyCyclespeed(bool direction, uint8_t dutycycle);
+        void    setHallBLDCmotorRPMspeed(bool direction, uint16_t desired_rpmSpeed, bool FieldWeakening);
```
Comment (jaenrig-ifx) [inline src/IFX007T-Motor-Control.h:77]: Are all these new API functions added to the keywords.txt? 
The setBiDirMotorSpeed() API has been changed, meaning, the major version needs to be changed. 
This needs to be changed then in the library.properties. 
Overloading the function would be possibility as well (if the previous function still makes sense to keep). Then  the next version could only be a minor increase.

File: keywords.txt [hunk 8]
```diff
@@ -19,6 +20,8 @@ setBiDirMotorSpeed	KEYWORD2
 configureBLDCMotor	KEYWORD2
 setBLDCmotorRPMspeed	KEYWORD2
 setBLDCDutyCyclespeed	KEYWORD2
+setHallBLDCmotorRPMspeed	KEYWORD2
```
Comment (jaenrig-ifx) [inline keywords.txt:23]: As mentioned... There might be some API functions still not added here?

**Review by jaenrig-ifx — CHANGES_REQUESTED** [review]
👯‍♀️ very nice Mr. @BrixInMotion.
The motor controlling functionalities I do not judge :) 
Only a few formalities...

File: src/IFX007T-Motor-Control.h (hunk 7 above)
Comment (BrixInMotion) [reply src/IFX007T-Motor-Control.h:77]: Ok I restored the old BiDirMotorSpeed function and the new one is now overloaded. Hardware tested again as well. I also inreased the Version number in the library.properties to 1.1.0 is this ok?

File: keywords.txt (hunk 8 above)
Comment (BrixInMotion) [reply keywords.txt:23]: One was missing yes, I included it as well.

**Comment by BrixInMotion** [comment]
Thanks for the reviews @jaenrig-ifx  and @djaumann 😀 !

**Review by jaenrig-ifx — APPROVED** [review]

**Pull request merged**